packages/docusaurus-theme-common/src/utils/useTOCHighlight.ts

```diff
@@ -138,19 +141,30 @@ function useTOCHighlight(config: TOCHighlightConfig | undefined): void {
       linkActiveClassName,
       minHeadingLevel,
       maxHeadingLevel,
+      autoScrollTOC,
     } = config;
 
-    function updateLinkActiveClass(link: HTMLAnchorElement, active: boolean) {
-      if (active) {
-        if (lastActiveLinkRef.current && lastActiveLinkRef.current !== link) {
-          lastActiveLinkRef.current?.classList.remove(linkActiveClassName);
-        }
-        link.classList.add(linkActiveClassName);
-        lastActiveLinkRef.current = link;
-        link.scrollIntoView({block: 'nearest'});
-      } else {
-        link.classList.remove(linkActiveClassName);
+    // When the page is scrolling (either a user scroll or smooth CSS scroll),
+    // We need to defer the TOC scroll or the two concurrent scrolls would block
+    // each other.
+    function scheduleScroll(link: HTMLAnchorElement) {
+      if (!autoScrollTOC) {
+        return () => {};
       }
+      const handle = setTimeout(() => {
+        const linkRect = link.getBoundingClientRect();
```

Comment:
all this code is quite low-level, please extract it under a meaningful function name that convey intent, and allowing us to more easily migrate to the official "if-needed" feature later?

```js
scrollIntoView(node, {
  scrollMode: 'if-needed',
  block: 'nearest', 
  behavior: 'smooth'
})
```

Maybe using this package? https://github.com/stipsan/scroll-into-view-if-needed

Reply:
This is not if-needed. It's the exact opposite: we don't scroll if we have to do a horizontal scroll. We don't care if we do a moot scroll, but we don't want the scroll to be horizontal.